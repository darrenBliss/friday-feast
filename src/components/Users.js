// @flow
import React, { Component, Fragment } from 'react'
import UserItem from '../components/UserItem'
import styled from 'styled-components'
import firebase from '../fire'
import { isEqual } from 'lodash'
import { Button, AltButton } from '../style/Button'
import { Card } from '../style/Card'
import { PlaceholderUsers } from '../style/PlaceholderUsers'

const SaveHolder = styled.div`
  background-color: #fff;
  padding: 1rem;
  margin-top: 1rem;

  button {
    margin-right: .2rem;
  }
`

const AddHolder = styled.form`
  display: flex;
  input {
    flex: 1 1 auto;
    margin-right: .2rem;
  }
`

type State = {
  value: string,
  users: Array<Object>,
  uid: number,
  activeUserIndex: ?number,
  isChanged: boolean,
  snapshot: Object
}

class Users extends Component<{}, State> {
  handleChange: () => void
  addUser: () => void
  onRemove: () => void
  handleUserEdit: () => void
  onMove: () => void
  saveData: () => void
  resetData: () => void
  setActiveUser: () => void

  constructor() {
    super()

    this.state = {
      value: '',
      users: [],
      uid: 0,
      activeUserIndex: null,
      isChanged: false,
      snapshot: {
        users: [],
        uid: null,
        activeUserIndex: null
      }
    }

    this.handleChange = this.handleChange.bind(this)
    this.addUser = this.addUser.bind(this)
    this.onRemove = this.onRemove.bind(this)
    this.handleUserEdit = this.handleUserEdit.bind(this)
    this.onMove = this.onMove.bind(this)
    this.saveData = this.saveData.bind(this)
    this.resetData = this.resetData.bind(this)
    this.setActiveUser = this.setActiveUser.bind(this)
  }

  handleChange(event: SyntheticInputEvent<HTMLInputElement>) {
    this.setState({
      value: event.currentTarget.value
    })
  }

  addUser(event: SyntheticEvent<>) {
    event.preventDefault()
    const trimmedVal: string = this.state.value.trim()

    const userObj = {
      name: trimmedVal,
      uid: this.state.uid
    }

    if (trimmedVal) {
      this.setState((prevState) => ({
        users: [...prevState.users, userObj],
        value: '',
        uid: prevState.uid + 1
      }))
    }
  }

  checkActiveUserSet = () => {
    // on delete, set active user to last item in user array if current active user is already the last item
    const usersLength = this.state.users.length
    const { activeUserIndex } = this.state

    this.setState(prevState => {
      if ((typeof prevState.activeUserIndex === 'number') && (usersLength <= prevState.activeUserIndex)) {
        return ({
          activeUserIndex: prevState.activeUserIndex -1
        })
      }
    })
  }

  onRemove(id: number) {
    this.setState(prevState => ({
      users: prevState.users.filter(item => item.uid !== id)
    }), () => {
      this.checkActiveUserSet()
    })
  }

  handleUserEdit(newText: string, id: number) {
    const index: number = this.state.users.findIndex(({ uid }) => uid === id)

    this.setState(prevState => ({
      users: [
        ...prevState.users.slice(0, index),
        {
          ...prevState.users[index],
          name: newText
        },
        ...prevState.users.slice(index + 1)
      ]
    }))
  }

  onMove(direction: number, currIndex: number) {
    const newIndex: number = currIndex + direction

    if (newIndex < 0 || newIndex >= this.state.users.length) {
      return
    }

    const excludedArr = this.state.users.filter((item, index) => index !== currIndex)

    const newArr = [
      ...excludedArr.slice(0, newIndex),
      {
        ...this.state.users[currIndex]
      },
      ...excludedArr.slice(newIndex)
    ]

    this.setState(prevState => ({
      users: newArr
    }))
  }

  componentDidMount() {
    //Firebase in here
    const arrFire = ['users', 'usersCurrentUID', 'activeUserKey']
    const arrState = ['users', 'uid', 'activeUserIndex']

    arrFire.forEach((item, index) => {
      firebase.database().ref(item).on('value', snapshot => {
        let newState = {}
        newState[arrState[index]] = snapshot.val()

        this.setState((prevState, props) => ({
          ...newState,
          snapshot: {
            ...prevState.snapshot,
            ...newState
          }
        }))
      })
    })
  }

  componentWillUnmount() {
    firebase.database().ref('users').off()
    firebase.database().ref('usersCurrentUID').off()
    firebase.database().ref('activeUserKey').off()
  }

  componentDidUpdate(prevProps: Object, prevState: Object) {
    const currState = {
      users: [...this.state.users],
      uid: this.state.uid,
      activeUserIndex: this.state.activeUserIndex
    }

    // Compare original state with current
    const didChanged = !isEqual(this.state.snapshot, currState)

    if (prevState.isChanged !== didChanged) {
      this.setState({
        isChanged: didChanged
      })
    }
  }

  saveData() {
    //Firebase in here
    const itemsRef = firebase.database().ref('users')
    const currUID = firebase.database().ref('usersCurrentUID')
    const activeUserKey = firebase.database().ref('activeUserKey')

    itemsRef.set(this.state.users)
    currUID.set(this.state.uid)
    activeUserKey.set(this.state.activeUserIndex)
  }

  resetData() {
    // Revert back to saved state from firebase
    if (!window.confirm('Are you sure you want to cancel your changes?')) {
      return
    }

    const { users, uid, activeUserIndex } = this.state.snapshot

    this.setState({
      users: [...users],
      uid,
      activeUserIndex
    })
  }

  setActiveUser(index: number) {
    this.setState({
      activeUserIndex: index
    })
  }

  render() {
    const userList = this.state.users.map((user, index) => {
      return (
        <li key={user.uid}>
          <UserItem
            name={user.name}
            index={index}
            uid={user.uid}
            isActive={this.state.activeUserIndex === index && true}
            onMove={this.onMove}
            userEdited={this.handleUserEdit}
            onRemove={this.onRemove}
            onActiveUser={this.setActiveUser}
          />
        </li>
      )
    })

    const ActiveUser = () => {
      if (typeof this.state.activeUserIndex === 'number' && this.state.users.length > this.state.activeUserIndex) {
        return (
          <p>The persons turn this week is {this.state.users[this.state.activeUserIndex].name}</p>
        )
      } else {
        return ''
      }
    }

    return (
      <div>

        {/* <hr/>
        <ActiveUser />
        <hr/> */}
        <Card>
          <AddHolder onSubmit={this.addUser}>
            <input name="user-name"
              value={this.state.value}
              onChange={this.handleChange}
              type="text" />
            <Button type="submit">Add person</Button>
          </AddHolder>
        </Card>

        {this.state.users.length > 0 ?
          <Fragment>
            <ul>{userList}</ul>
            <SaveHolder>
              <Button type="button"
                disabled={!this.state.isChanged}
                onClick={this.saveData}
              >Save</Button>
              <AltButton type="button"
                disabled={!this.state.isChanged}
                onClick={this.resetData}
              >Cancel</AltButton>
            </SaveHolder>
          </Fragment>

          :

          <PlaceholderUsers />
        }

      </div>
    )
  }
}

export default Users