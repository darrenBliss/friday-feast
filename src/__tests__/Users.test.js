import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import 'jest-styled-components'
import Users from '../components/Users'

describe('Should render Users component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Users />);
  });

  it('snapshot should work', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Should render 2 items', () => {
    wrapper.setState({
      users: [
        { name: 'Megatron', uid: 4 },
        { name: 'Jojo', uid: 6 }
      ]
    })

    expect(wrapper.find('UserItem')).toHaveLength(2)
  })

  it('Active user should highlight', () => {
    wrapper.setState({
      users: [
        { name: 'Megatron', uid: 4 },
        { name: 'Jojo', uid: 6 }
      ],
      uid: 7,
      activeUserIndex: 1
    })

    expect(wrapper.find('UserItem').at(1).props().isActive).toEqual(true)
  })

  it('Snapshot compare saveholder disabled prop should be false', () => {
    wrapper.setState({
      users: [
        { name: 'Megatron', uid: 4 },
        { name: 'Jojo', uid: 6 }
      ],
      uid: 7,
      activeUserIndex: 0,
      snapshot: {
        users: [{ name: 'Something new', uid: 2 }],
        uid: 7,
        activeUserIndex: null
      }
    })

    expect(wrapper.find('Users__SaveHolder Button').at(1).props().disabled).toEqual(false)
  })

  it('Snapshot deep compare saveholder disabled prop should be false', () => {
    wrapper.setState({
      users: [
        { name: 'Megatron', uid: 4 },
        { name: 'Jojo', uid: 6, test: { name: 'test', age: 4 } }
      ],
      uid: 7,
      activeUserIndex: 0,
      snapshot: {
        users: [
          { name: 'Megatron', uid: 4 },
          { name: 'Jojo', uid: 6, test: { name: 'test!', age: 5 } }
        ],
        uid: 7,
        activeUserIndex: 0
      }
    })

    expect(wrapper.find('Users__SaveHolder Button').at(1).props().disabled).toEqual(false)
  })

  it('Snapshot compare saveholder disabled prop should be true', () => {
    wrapper.setState({
      users: [
        { name: 'Megatron', uid: 4 }
      ],
      uid: 7,
      activeUserIndex: 0,
      snapshot: {
        users: [{ name: 'Megatron', uid: 4 }],
        uid: 7,
        activeUserIndex: 0
      }
    })

    expect(wrapper.find('Users__SaveHolder Button').at(1).props().disabled).toEqual(true)
  })

  it('Window confirm to be called', () => {
    window.confirm = jest.fn(() => true)

    const mountwrapper = mount(<Users />);

    mountwrapper.setState({
      users: [
        { name: 'Megatron', uid: 4 }
      ],
      uid: 7,
      activeUserIndex: 0,
      snapshot: {
        users: [{ name: 'Chopper', uid: 4 }],
        uid: 7,
        activeUserIndex: 0
      }
    })

    mountwrapper.find('Users__SaveHolder Button').at(1).simulate('click')

    expect(window.confirm).toBeCalled()
  })
})


