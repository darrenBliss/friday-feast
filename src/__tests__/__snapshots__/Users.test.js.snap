// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Users component addUser fn test snapshot 1`] = `
.c3 {
  display: inline-block;
  font-size: 1rem;
  color: #fff;
  padding: 1rem;
  background-color: #FF5A5F;
  border: 0;
  border-radius: 3px;
  opacity: 1;
}

.c6 {
  display: inline-block;
  font-size: 1rem;
  color: #fff;
  padding: 1rem;
  background-color: #FF5A5F;
  border: 0;
  border-radius: 3px;
  opacity: 1;
  color: #FF5A5F;
  border: 1px solid #FF5A5F;
  background-color: #fff;
  padding: 0 0.3rem;
}

.c9 {
  display: inline-block;
  font-size: 1rem;
  color: #fff;
  padding: 1rem;
  background-color: #FF5A5F;
  border: 0;
  border-radius: 3px;
  opacity: 1;
  color: #FF5A5F;
  border: 1px solid #FF5A5F;
  background-color: #fff;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 1rem;
  border-bottom: 1px solid #eee;
  background-color: #fff;
}

.c7 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: grid;
  grid-template-columns: 1fr repeat(3,auto);
  grid-column-gap: .2rem;
}

.c8 {
  padding: 1rem;
}

.c5 {
  display: grid;
  grid-row-gap: .2rem;
  margin-right: .2rem;
}

.c1 {
  background-color: #fff;
  padding: 1rem;
  box-shadow: 0 0 4px rgba(0,0,0,.3);
  border-radius: 3px;
  max-width: 400px;
  margin: 2rem auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 input {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  margin-right: .2rem;
}

.c0 {
  width: 100%;
  max-width: 600px;
  margin: 2rem auto;
}

.c10 {
  background-color: #fff;
  padding: 1rem;
  margin-top: 1rem;
}

.c10 button {
  margin-right: .2rem;
}

<Users>
  <Users__Page>
    <div
      className="c0"
    >
      <AddUser
        onAddUser={[Function]}
      >
        <Card>
          <div
            className="c1"
          >
            <AddUser__AddHolder
              onSubmit={[Function]}
            >
              <form
                className="c2"
                onSubmit={[Function]}
              >
                <input
                  name="user-name"
                  onChange={[Function]}
                  type="text"
                  value=""
                />
                <Button
                  type="submit"
                >
                  <button
                    className="c3"
                    type="submit"
                  >
                    Add person
                  </button>
                </Button>
              </form>
            </AddUser__AddHolder>
          </div>
        </Card>
      </AddUser>
      <ul>
        <li
          key="4"
        >
          <UserItem
            index={0}
            isActive={false}
            name="Megatron"
            onActiveUser={[Function]}
            onMove={[Function]}
            onRemove={[Function]}
            uid={4}
            userEdited={[Function]}
          >
            <UserItem__Holder
              isActive={false}
            >
              <div
                className="c4"
              >
                <UserItem__DirectionButtons>
                  <div
                    className="c5"
                  >
                    <Button
                      onClick={[Function]}
                      title="Up"
                      type="button"
                    >
                      <button
                        className="c6"
                        onClick={[Function]}
                        title="Up"
                        type="button"
                      >
                        ↑
                      </button>
                    </Button>
                    <Button
                      onClick={[Function]}
                      title="Down"
                      type="button"
                    >
                      <button
                        className="c6"
                        onClick={[Function]}
                        title="Down"
                        type="button"
                      >
                        ↓
                      </button>
                    </Button>
                  </div>
                </UserItem__DirectionButtons>
                <UserItem__ListBox>
                  <div
                    className="c7"
                  >
                    <UserItem__ItemText>
                      <span
                        className="c8"
                      >
                        Megatron
                      </span>
                    </UserItem__ItemText>
                    <Button
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c3"
                        onClick={[Function]}
                        type="button"
                      >
                        Edit
                      </button>
                    </Button>
                    <Button
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c3"
                        onClick={[Function]}
                        type="button"
                      >
                        Delete
                      </button>
                    </Button>
                    <Button
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c9"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Set as active
                      </button>
                    </Button>
                  </div>
                </UserItem__ListBox>
              </div>
            </UserItem__Holder>
          </UserItem>
        </li>
        <li
          key="6"
        >
          <UserItem
            index={1}
            isActive={false}
            name="GG"
            onActiveUser={[Function]}
            onMove={[Function]}
            onRemove={[Function]}
            uid={6}
            userEdited={[Function]}
          >
            <UserItem__Holder
              isActive={false}
            >
              <div
                className="c4"
              >
                <UserItem__DirectionButtons>
                  <div
                    className="c5"
                  >
                    <Button
                      onClick={[Function]}
                      title="Up"
                      type="button"
                    >
                      <button
                        className="c6"
                        onClick={[Function]}
                        title="Up"
                        type="button"
                      >
                        ↑
                      </button>
                    </Button>
                    <Button
                      onClick={[Function]}
                      title="Down"
                      type="button"
                    >
                      <button
                        className="c6"
                        onClick={[Function]}
                        title="Down"
                        type="button"
                      >
                        ↓
                      </button>
                    </Button>
                  </div>
                </UserItem__DirectionButtons>
                <UserItem__ListBox>
                  <div
                    className="c7"
                  >
                    <UserItem__ItemText>
                      <span
                        className="c8"
                      >
                        GG
                      </span>
                    </UserItem__ItemText>
                    <Button
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c3"
                        onClick={[Function]}
                        type="button"
                      >
                        Edit
                      </button>
                    </Button>
                    <Button
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c3"
                        onClick={[Function]}
                        type="button"
                      >
                        Delete
                      </button>
                    </Button>
                    <Button
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c9"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Set as active
                      </button>
                    </Button>
                  </div>
                </UserItem__ListBox>
              </div>
            </UserItem__Holder>
          </UserItem>
        </li>
        <li
          key="7"
        >
          <UserItem
            index={2}
            isActive={false}
            name="totoro"
            onActiveUser={[Function]}
            onMove={[Function]}
            onRemove={[Function]}
            uid={7}
            userEdited={[Function]}
          >
            <UserItem__Holder
              isActive={false}
            >
              <div
                className="c4"
              >
                <UserItem__DirectionButtons>
                  <div
                    className="c5"
                  >
                    <Button
                      onClick={[Function]}
                      title="Up"
                      type="button"
                    >
                      <button
                        className="c6"
                        onClick={[Function]}
                        title="Up"
                        type="button"
                      >
                        ↑
                      </button>
                    </Button>
                    <Button
                      onClick={[Function]}
                      title="Down"
                      type="button"
                    >
                      <button
                        className="c6"
                        onClick={[Function]}
                        title="Down"
                        type="button"
                      >
                        ↓
                      </button>
                    </Button>
                  </div>
                </UserItem__DirectionButtons>
                <UserItem__ListBox>
                  <div
                    className="c7"
                  >
                    <UserItem__ItemText>
                      <span
                        className="c8"
                      >
                        totoro
                      </span>
                    </UserItem__ItemText>
                    <Button
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c3"
                        onClick={[Function]}
                        type="button"
                      >
                        Edit
                      </button>
                    </Button>
                    <Button
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c3"
                        onClick={[Function]}
                        type="button"
                      >
                        Delete
                      </button>
                    </Button>
                    <Button
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      <button
                        className="c9"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Set as active
                      </button>
                    </Button>
                  </div>
                </UserItem__ListBox>
              </div>
            </UserItem__Holder>
          </UserItem>
        </li>
      </ul>
      <Users__SaveHolder>
        <div
          className="c10"
        >
          <Button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <button
              className="c3"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Save
            </button>
          </Button>
          <Button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <button
              className="c9"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </Button>
        </div>
      </Users__SaveHolder>
    </div>
  </Users__Page>
</Users>
`;

exports[`Should render Users component snapshot should work 1`] = `
<Users__Page>
  <AddUser
    onAddUser={[Function]}
  />
  <PlaceholderUsers />
  <Users__SaveHolder>
    <Button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Save
    </Button>
    <Button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Cancel
    </Button>
  </Users__SaveHolder>
</Users__Page>
`;
